cookbook/case_studies/ml_training/mnist_classifier/pytorch_single_node_multi_gpu.py

```diff
@@ -0,0 +1,405 @@
+    accuracies = []
+    for epoch in range(1, hp.epochs + 1):
+        train(model, rank, training_data_loader, optimizer, epoch, hp.log_interval)
+
+        # only compute validation metrics in the main process
+        if rank == 0:
+            accuracies.append(test(model, rank, test_data_loader))
+        dist.barrier()  # wait for main process to complete validation before continuing training
+
+    if rank == 0:
+        wandb.finish()  # this is important to tell to wandb that we're done logging metrics
+
+        # after training the model, we can simply save it to disk and return it from the Flyte task as a :py:class:`flytekit.types.file.FlyteFile`
+        # type, which is the ``PythonPickledFile``. ``PythonPickledFile`` is simply a decorator on the ``FlyteFile`` that records the format
+        # of the serialized model as ``pickled``
+        print("Saving model")
+        torch.save(model.state_dict(), MODEL_FILE)
+
+        # save epoch accuracies
+        print("Saving accuracies")
+        with open(ACCURACIES_FILE, "w") as fp:
+            json.dump(accuracies, fp)
+
+    print(f"Rank {rank + 1}/{world_size} process complete.\n")
+    dist.destroy_process_group()  # clean up
+
+
+# The output model using :py:func:`pytorch:torch.save` saves the `state_dict` as described
+# `in pytorch docs <https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-and-loading-models>`_.
+# A common convention is to have the ``.pt`` extension for the model file.
+#
+# .. note::
+#    Note the usage of ``requests=Resources(gpu=WORLD_SIZE)``. This will force Flyte to allocate this task onto a
+#    machine with GPU(s), which in our case is 2 gpus. The task will be queued up until a machine with GPU(s) can be
+#    procured. Also, for the GPU Training to work, the Dockerfile needs to be built as explained in the
+#    :ref:`pytorch-dockerfile` section.
+
+@task(
+    retries=2,
+    cache=True,
+    cache_version="1.2",
+    requests=Resources(gpu=str(WORLD_SIZE), mem="30Gi", storage="20Gi", ephemeral_storage="500Mi"),
+    limits=Resources(gpu=str(WORLD_SIZE), mem="30Gi", storage="20Gi", ephemeral_storage="500Mi"),
+)
+def pytorch_mnist_task(hp: Hyperparameters) -> TrainingOutputs:
+    print("Start MNIST training:")
+
+    world_size = torch.cuda.device_count()
+    print(f"Device count: {world_size}")
+    download_mnist(DATA_DIR)
+    mp.spawn(
+        train_mnist,
+        args=(world_size, hp),
+        nprocs=world_size,
+        join=True,
+    )
+    print("Training Complete")
+    with open(ACCURACIES_FILE) as fp:
+        accuracies = json.load(fp)
+    return TrainingOutputs(
+        epoch_accuracies=accuracies, model_state=PythonPickledFile(MODEL_FILE)
+    )
+
+
+
+# %%
+# Finally, we define a workflow to run the training algorithm. We return the model and accuracies.
+@workflow
+def pytorch_training_wf(hp: Hyperparameters = Hyperparameters(epochs=10, batch_size=128)) -> TrainingOutputs:
```

Comment:
should we also add some common task at the end, that maybe calculates the accuracy of a validation set?

Reply:
`TrainingOutputs` already contains the validation accuracies, so not sure what task makes sense, since wandb is already gathering all the metrics.